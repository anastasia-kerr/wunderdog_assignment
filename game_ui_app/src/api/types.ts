import { Gestures } from '@/types'

type BaseApiResponse<T> = {
    succeeded: boolean;
    errors: any[];
    result: T;
}
export type CreateGameResponse = BaseApiResponse<{
    id: number
}>
export type JoinCurrentRoundResponse = BaseApiResponse<{
    id: number;
    roundNumber: number
}>
export type GetCurrentRoundPlayersResponse = BaseApiResponse<{
    players: string[]
}>
export type RoundResult = {
    isDraw?: boolean;
    isLastRound?: boolean;
    loserGesture?: Gestures;
    loserNickname?: string;
    winnerGesture?: Gestures;
    winnerNickname?: string;
}
export type GetRoundResultResponse = BaseApiResponse<RoundResult>
export type GetGameResultResponse = BaseApiResponse<
{winnerNickname?: string}
>
export type LeaveCurrentRoundResponse = JoinCurrentRoundResponse
export type PlaceGestureResponse = JoinCurrentRoundResponse
