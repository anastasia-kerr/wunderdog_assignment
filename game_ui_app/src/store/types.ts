import { RoundResult } from '@/api/types'

export interface RootState {
    error: string,
    gameId: string,
    roundNumber: number,
    currentPlayers: string[],
    gameWinner: string,
    nickname?: string|null,
    playerIdentifier?: string|null,
    roundResults: RoundResult
  }
