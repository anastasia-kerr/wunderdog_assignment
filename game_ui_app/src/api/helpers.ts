import config from '../configs'

export const poll = async (callback:() => Promise<any>) => {
  let pollCount = 0
  const handler = async () => {
    pollCount++
    const result = await callback()
    if (result || pollCount === config.maxPollingTimeInSeconds) {
      clearInterval(intervalId)
    }
  }
  const intervalId = setInterval(handler, config.pollingIntervalInMs)
}
