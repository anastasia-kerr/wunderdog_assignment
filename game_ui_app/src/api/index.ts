import axios from 'axios'
import { CreateGameResponse, GetCurrentRoundPlayersResponse, GetGameResultResponse, GetRoundResultResponse, JoinCurrentRoundResponse, LeaveCurrentRoundResponse, PlaceGestureResponse } from './types'

function setUpInterceptors () {
  axios.interceptors.response.use(null, (error) => {
    if (error?.response?.data?.Errors) {
      console.log('Error while calling the api: ', JSON.stringify(error?.response?.data?.Errors))
    } else {
      console.log('unexpected error while calling the api: ', error)
    }
    return Promise.reject(error)
  })
}
export function setApiDefaults (url:string) {
  axios.defaults.baseURL = url
  setUpInterceptors()
}
export async function createGame (payload : {nickname:string, numberOfRounds:number}):Promise<number> {
  const { data } = await axios.post<CreateGameResponse>(
    'api/Games',
    payload
  )
  return data.result.id
}
export async function joinCurrentRound (payload : { gameId:string, playerNickname:string, playerIdentifier:string}):Promise<number> {
  const { data } = await axios.put<JoinCurrentRoundResponse>(
    `api/Games/${payload.gameId}/join`,
    payload
  )
  return data.result.roundNumber
}
export async function getCurrentRoundPlayers (gameId:string):Promise<string[]> {
  const { data } = await axios.get<GetCurrentRoundPlayersResponse>(
    `api/Games/${gameId}/players`
  )
  return data.result.players
}
export async function getRoundResults (gameId:string, round:number):Promise<GetRoundResultResponse> {
  const { data } = await axios.get<GetRoundResultResponse>(
    `api/Games/${gameId}/result/${round}`
  )
  return data
}
export async function getGameResult (gameId:string):Promise<GetGameResultResponse> {
  const { data } = await axios.get<GetGameResultResponse>(
    `api/Games/${gameId}/result/`
  )
  return data
}
export async function leaveCurrentRound (payload : { gameId:string, playerIdentifier:string}):Promise<number> {
  const { data } = await axios.put<LeaveCurrentRoundResponse>(
    `api/Games/${payload.gameId}/leave`,
    payload
  )
  return data.result.roundNumber
}
export async function placeGesture (payload : { gameId:string, playerNickname:string, playerIdentifier:string, roundGesture:number}):Promise<number> {
  const { data } = await axios.post<PlaceGestureResponse>(
    `api/Games/${payload.gameId}/gesture`,
    payload
  )
  return data.result.id
}
