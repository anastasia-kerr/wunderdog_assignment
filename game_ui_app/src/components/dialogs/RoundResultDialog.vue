<template>
  <v-dialog :width="dialogSize" persistent v-model="dialog">
    <v-card v-if="showLoader" :loading="true">
      <v-card-text>
        <div class="mt-2">{{ waitingLbl }}</div>
      </v-card-text>
    </v-card>
    <v-card v-else>
      <v-card-title v-text="headerRoundLabel" />
      <v-card-text>
        <v-row>
          <v-col> {{ roundResultMessage }} </v-col>
        </v-row>
      </v-card-text>
        <v-card-title v-if="isLastRound" v-text="headerGameLabel" />
        <v-card-text v-if="isLastRound">
          <v-row>
            <v-col> {{ gameResultMessage }} </v-col>
          </v-row>
        </v-card-text>
      <v-card-actions>
        <v-spacer />
        <v-btn color="primary" v-text="btnLabel" @click="onBtnPress" />
      </v-card-actions>
    </v-card>
  </v-dialog>
</template>

<script lang="ts">
import Vue from 'vue'
import { mapGetters, mapState } from 'vuex'
import config from '../../configs'
const { smallDialogSize } = config

export default Vue.extend({
  name: 'RoundResult',
  data: () => ({
    dialog: true,
    dialogSize: smallDialogSize,
    headerRoundLabel: 'Round results',
    headerGameLabel: 'Game results'
  }),
  computed: {
    ...mapGetters(['opponent', 'roundResultMessage', 'gameResultMessage']),
    ...mapState(['roundResults']),
    showLoader (): boolean {
      return !this.roundResults.id
    },
    waitingLbl (): string {
      return this.opponent
        ? `Waiting for ${this.opponent} to make their move`
        : 'Waiting for an opponent to join the game'
    },
    isLastRound (): boolean {
      return this.roundResults.isLastRound === true
    },
    btnLabel (): string {
      return this.isLastRound ? 'Start new game' : 'Play another round'
    }
  },
  watch: {
    opponent: {
      async handler (newValue) {
        if (!newValue) {
          await this.$store.dispatch('setCurrentPlayers')
        } else {
          await this.$store.dispatch('setCurrentRoundResults')
        }
      },
      immediate: true
    }
  },
  methods: {
    async onBtnPress (): Promise<void> {
      if (this.isLastRound) {
        this.$router.push({ name: 'create-game' })
      } else {
        await this.$store.dispatch('joinCurrentRound')
        this.dialog = false
        this.$emit('closed')
      }
    }
  }
})
</script>
