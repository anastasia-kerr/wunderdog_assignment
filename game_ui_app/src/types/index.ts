export enum Gestures {
    Rock = 0,
    Paper = 1,
    Scissors = 2
  }
