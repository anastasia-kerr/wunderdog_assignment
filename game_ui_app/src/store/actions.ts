import { ActionTree } from 'vuex'
import { RootState } from './types'
import { leaveCurrentRound, joinCurrentRound, getCurrentRoundPlayers, getRoundResults, getGameResult, placeGesture } from '@/api'
import { poll } from '@/api/helpers'

const actions: ActionTree<RootState, RootState> = {
  setUser (context, nickname: string) {
    context.commit('setNickname', nickname)
    localStorage.setItem('nickname', nickname ?? '')
    localStorage.setItem('playerIdentifier', nickname ? nickname + Date.now().toString() : '')
  },
  async leaveGame ({ state, dispatch }) {
    if (state.playerIdentifier) {
      await leaveCurrentRound({ gameId: state.gameId, playerIdentifier: state.playerIdentifier })
    }
    dispatch('setUser', '')
  },
  async joinCurrentRound ({ state, commit, dispatch }) {
    if (!state.playerIdentifier || !state.nickname) {
      throw new Error('Player must have some identifier or name')
    }

    try {
      const roundNumber = await joinCurrentRound({
        gameId: state.gameId,
        playerNickname: state.nickname,
        playerIdentifier: state.playerIdentifier
      })

      commit('resetGameState')
      commit('setRoundNumber', roundNumber)

      dispatch('setCurrentPlayers')
    } catch (error : any) {
      commit('setError', error?.response?.data?.Errors)
    }
  },
  async setCurrentPlayers ({ commit, state }) {
    return await poll(async () => {
      const result = await getCurrentRoundPlayers(state.gameId)
      if (result.length === 2) {
        commit('setCurrentPlayers', result)
        return true
      } else if (state.currentPlayers.length === 2) {
        return true
      }
      return false
    })
  },
  async setCurrentRoundResults ({ state, commit, dispatch }) {
    return await poll(async () => {
      const response = await getRoundResults(state.gameId, state.roundNumber)
      if (response.succeeded) {
        commit('setCurrentRoundResults', response.result)
        if (response.result.isLastRound) {
          dispatch('setGameResult')
        }
        return true
      }
      return false
    })
  },
  async setGameResult ({ commit, state }) {
    const response = await getGameResult(state.gameId)
    commit('setGameResult', response.result.winnerNickname)
  },
  async placeGesture (context, payload: { gesture: number }) {
    if (context.state.playerIdentifier && context.state.nickname) {
      await placeGesture({
        gameId: context.state.gameId,
        playerNickname: context.state.nickname,
        playerIdentifier: context.state.playerIdentifier,
        roundGesture: payload.gesture
      })
    }
  }
}
export default actions
