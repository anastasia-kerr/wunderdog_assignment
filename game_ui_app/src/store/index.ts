import { RoundResult } from '@/api/types'
import Vue from 'vue'
import Vuex from 'vuex'
import { RootState } from './types'
import getters from './getters'
import mutations from './mutations'
import actions from './actions'

Vue.use(Vuex)
export const state: RootState = {
  gameId: '',
  error: '',
  roundNumber: 0,
  currentPlayers: [] as string[],
  gameWinner: '',
  nickname: localStorage.getItem('nickname'),
  playerIdentifier: localStorage.getItem('playerIdentifier'),
  roundResults: {} as RoundResult
}

export default new Vuex.Store({
  state,
  getters,
  mutations,
  actions
})
