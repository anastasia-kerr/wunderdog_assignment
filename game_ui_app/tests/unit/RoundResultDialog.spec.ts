import { RoundResult } from '@/api/types'
import Vue from 'vue'
import RoundResultDialog from '@/components/dialogs/RoundResultDialog.vue'
import { RootState } from '@/store/types'
import { mount, createLocalVue, Wrapper } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vuex, { Store } from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('RoundResultDialog', () => {
  let store: Store<RootState>
  let vuetify: Vuetify
  let wrapper:Wrapper<Vue>
  const player1 = 'player1'
  const player2 = 'player2'
  const roundNumber = 1
  const state = {
    gameId: '1',
    error: '',
    roundNumber: roundNumber,
    currentPlayers: [player1, player2],
    gameWinner: '',
    nickname: player1,
    playerIdentifier: player1 + 'playerIdentifier',
    roundResults: { id: 1, isLastRound: true } as RoundResult
  }
  let actions
  let getters
  const roundResultMessage = 'roundResultMessage'
  const gameResultMessage = 'gameResultMessage'
  beforeEach(() => {
    vuetify = new Vuetify()
    actions = {
      setCurrentRoundResults: jest.fn(),
      setCurrentPlayers: jest.fn()
    }
    getters = {
      opponent: () => player2,
      roundResultMessage: () => roundResultMessage,
      gameResultMessage: () => gameResultMessage
    }
    store = new Vuex.Store({
      state: { ...state },
      actions,
      getters
    })

    wrapper = mount(RoundResultDialog, { store, localVue, vuetify })
  })
  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it('Round is finsihed Must Display roundResultMessage', () => {
    expect(wrapper.text()).toContain(roundResultMessage)
  })
  it('Game is finsihed Must Display gameResultMessage', () => {
    expect(wrapper.text()).toContain(gameResultMessage)
  })
})
