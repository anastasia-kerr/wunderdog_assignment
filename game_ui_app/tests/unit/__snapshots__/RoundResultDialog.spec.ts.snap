// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoundResultDialog renders correctly 1`] = `
<v-dialog
  persistent=""
  width="320"
>
  <v-card>
    <v-card-title>
      Round results
    </v-card-title>
     
    <v-card-text>
      <v-row>
        <v-col>
           roundResultMessage 
        </v-col>
      </v-row>
    </v-card-text>
     
    <v-card-title>
      Game results
    </v-card-title>
     
    <v-card-text>
      <v-row>
        <v-col>
           gameResultMessage 
        </v-col>
      </v-row>
    </v-card-text>
     
    <v-card-actions>
      <v-spacer />
       
      <v-btn
        color="primary"
      >
        Start new game
      </v-btn>
    </v-card-actions>
  </v-card>
</v-dialog>
`;
